### DISCUSSION TITLE
Run php-cs-fixer from docker

### DISCUSSION THREAD

**Discussion by zordsdavini** [phpactor/phpactor#2359] [General]
Can I run php-cs-fixer from docker? I have some issues like fixer cache has been rebuilt each time I edit any file. I have pre-commit php-cs-fixer git action what use docker. I want to use same environment for phpactor actions

**Comment by cizordj** [comment]
You can create a wrapper script that runs phpactor in the same container as your project runs, this way you can share the same cache files and all.

https://github.com/cizordj/yancfs/blob/master/bin/phpactor-wrapper.sh

**Comment by dantleech** [comment]
:point_up: creating a wrapper is the best way atm.

**Comment by zordsdavini** [comment]
But that means it should be in this container? 
Can't php-cs-fixer just be called with some kind of docker prefix to format? Ex. `docker exec -u nobody container_name /bin/bash -c "PHP_CS_FIXER_IGNORE_ENV=1 bin/php-cs-fixer fix"`

**Comment by zordsdavini** [comment]
tested more. 
One thing is different php versions on local and on docker (8.2.10 vs 8.2.8). This call's to regenerate cache because of "different" versions. 
Other thing is rules. In our job we use config file. Phpactor use it's own rules or even it calls them separately. Not sure why. So, different set of rules will regenerate cache again. For one file that's not so important, but for our project with 5k files cache is important. Regeneration takes 5-6 minutes.
For now I hardcoded config path and bumped php to same version but arch is very fast with upgrades, so it is temporary.
Will you accept PR with:
language_server_php_cs_fixer.config_path = 'bla/bla/bla'
language_server_php_cs_fixer.wrapper = 'docker exec -u nobody container_name /bin/bash -c'
If so, I will work on that

**Reply by dantleech** [reply]
Phpactor doesn't use it's own rules or config, it will use the default the cs fixer uses. Do you have a nonstandard config file? But if cs fixer let's you specify a custom one then yes, we can support that.

As for the wrapper... the bigger picture is that we should support this for other integrations too, so I'm thinking if there is a generic way to do it. The work around is to wrap it in a shell script as @cizordj  points out

**Reply by cizordj** [reply]
With the wrapper you can be sure that your project and phpactor are on the same page, same PHP version, same cs fixer and so on.

**Comment by zordsdavini** [comment]
Added PR to support custom config https://github.com/phpactor/phpactor/pull/2384
Today I got php update on Archlinux, so after this apply I will have problems (in project docker I have different version). I don't understand cizordj wrapper. I can't use phpactor in projects docker, so I'll try to give PR for my vision how to solve this

**Reply by cizordj** [reply]
Why not? You're not installing phpactor in your project, you're just using your project as the runtime environment. All analysis like _phpstan_, _php-cs-fixer_ and so on run on the same PHP container as the project itself. I know the script is complicated, it took me months to figure it out, I had the same problem as you have and instead of changing how phpactor works I decided to take the other path. 😃

**Reply by zordsdavini** [reply]
I'm not the only developer of my big project and not the main developer :) Other thing is my vim (neovim) configuration. For now I use it with Mason and phpactor fits very nice in it and all these analysis I'd like to leave in project level. It should be like:
- oh, it's php, let's use phpactor
- oh, it has some configuration - let's use it

Still have problems here - already half of year didn't figure out how to solve this and after phpactor update I don't have LSP actions but to lazy to fix it - on git commit project runs php-cs-fixer and phpstan

**Comment by zordsdavini** [comment]
Correct me if I wrong.
To create docker wrapper for php-cs-fixer we should do steps:
- add config `language_server_php_cs_fixer.wrapper = 'docker exec -u nobody container_name /bin/bash -c'`
- in `PhpCsFixerProcess.run()` on config enabled we should get formatted command and create new process wrapped in `""` and prefixed with wrapper from config

Maybe better to add wrap method in ProcessBuilder and call it from PhpCsFixerProcess during run to build everything?

**Comment by zordsdavini** [comment]
Added PR to support wrapper https://github.com/phpactor/phpactor/pull/2388
